### DISCUSSION TITLE
Otterkit COBOL Dev Update 2: Syntactic analysis is now complete

### DISCUSSION THREAD

**Discussion by KTSnowy** [otterkit/otterkit-cobol#18] [Announcements]
# Syntactic analysis is now complete:

After ~11500 lines of handwritten recursive descent, syntactic analysis on the parser is now complete. This means that the parser is now, grammatically and structurally speaking, complete! We'll now work on implementing semantic analysis which includes error checking, error recovery, and identifier name resolution. 

Most of the infrastructure needed for name resolution (the symbol table and related types) has already been implemented, and is now just a matter of refining it and putting everything together. Error recovery has also being implemented using anchor points, but we still need to find the best places to use them.

Fun fact, Otterkit now has (as far as we know) the one and only handwritten parser for COBOL 2023, and we're committed to work on making it the greatest Standard COBOL parser on the market. 

Making the parser fully handwritten means that we're not limited by a parser generator and its provided functionality, or the (sometimes) not very helpful error messages. We're able to fine tune absolutely everything down to exactly when and where to show an error to the user, including the contents of the error.

And while it is an enormous amount of work to write a handwritten recursive descent parser for COBOL, we still strongly believe that this is the right choice for Otterkit.

# Standard COBOL syntax checklist:

So which parts of Standard COBOL will we be supporting? In summary, we plan to support the whole standard, but we'll leave here a checklist of things that the parser "understands" syntactically speaking. This might be useful to help distinguish dialect COBOL from Standard COBOL because of how different (and sometimes incompatible) some third-party dialects are from the standard.

If you find that the parser incorrectly parses one of these (or crashes), please open an issue on this repo so that we can fix it 

(Note: Name resolution is WIP so some errors are expected).

### Identification division:

- Program-id
  - Prototypes
  - Common, initial and recursive options
- Function-id
  - Prototypes
- Class-id
  - Single class inheritance
  - Final (sealed) classes
  - Generic classes
- Interface-id
  - Multiple interface inheritance
  - Final (sealed) interfaces
  - Generic interfaces
- Factory
  - Multiple interface implements
- Object
  - Multiple interface implements
- Method-id
  - Override
  - Final (sealed) methods
- Method-id (property syntax)
  - Getter property
  - Setter property
  - Auto implemented properties
- Options paragraph
  - intermediate rounding,
  - default rounded, 
  - entry convention, 
  - initialize,
  - and arithmetic clauses

### Environment division:

- Configuration section
  - Source-computer
  - Object-computer 
  - Special-names
  - Repository
- Input-output section
  - File-control
  - I-o-control

### Data division:

- File section 
  - File entries
  - Record entries
  - Typedef entries
  - Constant entries
- Working-storage, local-storage and linkage sections
  - 77-level entries
  - Data entries
  - Typedef entries
  - Constant entries
- Report section
  - Report entries
  - Report group entries
  - Constant entries
- Screen section
  - Screen entries
  - Constant entries

### Procedure Division:

- Procedure header
  - Without sections
  - With sections
    - With declarative error handling
  - Object oriented header
    - For interfaces and classes
  - With raising exception phrase
- Supported statements
  - (All 51 standard statements)
  - Accept statement
  - Add statement
  - Allocate statement 
    - (similar to C malloc)
  - Call statement
  - Cancel statement
  - Close statement
  - Commit statement
  - Compute statement
  - Continue statement
  - Delete statement
  - Display statement
  - Divide statement
  - Evaluate statement
  - Exit statement
  - Free statement 
    - (similar to C free)
  - Generate statement
  - Goback statement
  - Go statement
    - (go to)
  - If statement
  - Initialize statement
  - Initiate statement
  - Inspect statement
  - Invoke statement
    - (For method calls)
    - (Inline calls are also supported)
  - Merge statement
  - Move statement
  - Multiply statement
  - Open statement
  - Perform statement
  - Raise statement
  - Read statement
  - Receive statement
    - (New async messaging)
  - Release statement
  - Resume statement
  - Return statement
  - Rewrite statement
  - Rollback statement
  - Search statement
  - Send statement
    - (New async messaging)
  - Set statement
    - (17 formats)
  - Sort statement
  - Start statement
  - Stop statement 
    - (stop run)
  - String statement
  - Subtract statement
  - Suppress statement
  - Terminate statement
  - Unlock statement
  - Unstring statement
  - Use statement
  - Validate statement
  - Write statement

We'll also be supporting all intrinsic functions from the COBOL 2023 standard, but we need name resolution working to correctly parse their arguments (some of which contain reserved words).

# New error message API:

We've overhauled our old error messages. This new error message system not only looks much better, but is also much easier to use compared to the previous one.

Here's an example of how the new error messages look:
![227745371-86b6ee99-fc5e-491f-9d83-5b756b599fa8](https://user-images.githubusercontent.com/73310001/231919160-c5c21156-7a9b-4738-a1f3-a1ed23b344fb.png)

And here's an example of the ease of use:
![Screenshot 2023-04-13 224145](https://user-images.githubusercontent.com/73310001/231919679-a15a1bb0-bf1c-4be3-9f09-0f02181576cb.png)

This error message system gives us very precise control over exactly where (in which token) and when to display a message back to the user, as well as which messages to show. It allows us to write helpful error messages that can help even a new developer slowly learn how write COBOL.

You might have also noticed that we're not passing any file information to the error message methods, that's because internally each token keeps track of the file from which it came from. This allows us to know exactly from which file, on which line and which column the token was written in, so that we can correctly display that information back to the user.

Each token keeping track of which file it came from means that Otterkit will always be able to correctly display that information back, no matter how nested your copybook imports are.

# Standard decimal math library:

To improve maintainability and reusability we moved our decimal math implementation from the runtime library into its own separate library.

The new package is called [Otterkit.Numerics](https://github.com/otterkit/otterkit-numerics) and is currently available on NuGet.

This library contains C# bindings around a software implementation of the IEEE 754 Decimal128 and Decimal64 types written in C. We're also planning on writing bindings for the smaller Decimal32 type.

# Next steps:

Now we'll start working on the semantic analysis stage of the parser with name resolution, and also on improving the current error messages and writing new ones. 

We're committed on working to make Otterkit have the greatest open source Standard COBOL parser on the market, and to continue working on the compiler in order to get it production-ready. We'll be sharing more information about Otterkit's development in the next dev update when we get semantic analysis working.

**-- Otterkit Project Team**

**Comment by GitMensch** [comment]
Note: for testing `DATA DIVISION` (variables in general, especially related to the field structure and `PICTURE`) I highly suggest to check NIST85 NC126A:

> PROGRAM NC126A TESTS THE USE OF LEVEL NUMBERS 01 THROUGH 49 INCLUSIVE WITH A VARIETY OF PICTURE CLAUSES AND GROUP AND ELEMENTARY COMPARISONS.

... here's a COBOL2014 version of it
[NC126A2.CBL](https://github.com/otterkit/otterkit/files/11531157/NC126A2.CBL.txt)

**Reply by KTSnowy** [reply]
Hey Simon, we might have to update the tests for this as well, because COBOL 2023 added this feature: 

"_User-defined PICTURE clause editing using the EDITING phrase of the PICTURE clause_".

The COBOL standard gives the following usage example of it:
```COBOL
01 item PIC IS L999.99 EDITING "L" FOR NEGATIVE IS "DEBIT " BLANK WHEN ZERO.
. . .

*> Result: DEBIT 123.45
MOVE -123.45 TO item.
```

This feature is incredibly powerful for more advanced formatting scenarios. This makes Standard COBOL have the easiest string formatting of any language I've seen so far.

It would also be amazing to have a copybook with a group of common formatting variables already defined. 

Otterkit will definitely **HAVE** to support this, we can't let the COBOL Working Group define another amazing feature as "obsolete due to no interest" again in the next standard (like the VALIDATE and binary arithmetic).

I say this because I can already see that major vendors might choose to not support it because legacy customers won't use it.

**Reply by GitMensch** [reply]
So... does otterkit currently pass NC126A2? Do you like its codegen for these record structures?

That should be verified first and yes - a test for the syntax (and later for that nice editing feature) should be added afterwards, too.

**Reply by KTSnowy** [reply]
The codegen currently doesn't support most of the syntax, I'm still working on name resolution and type checking. After that I'll start working on the codegen.

Also, interestingly enough the file above does not conform to the COBOL standard. The separator periods on it don't have a space after them, so Otterkit displays a TON of these errors for the whole file:

```
$> otterkit build -e .\NC126A2.cob --parse

 Analyzer Error [COB0025]: Data item definition, missing separator period.
     ╭─/> [NC126A2.cob:39:30]
     │
  39 │  01  PRINT-REC PICTURE X(120).
     │                        ~~~~~~~/> Expected a separator period '. ' after this token
     │
     │  Note: Every item must end with a separator period
     │
 ────╯

 Analyzer Error [COB0025]: Data item definition, missing separator period.
     ╭─/> [NC126A2.cob:40:33]
     │
  40 │  01  DUMMY-RECORD PICTURE X(120).
     │                           ~~~~~~~/> Expected a separator period '. ' after this token        
     │
     │  Note: Every item must end with a separator period
     │
 ────╯

 Analyzer Error [COB0025]: Data item definition, missing separator period.
     ╭─/> [NC126A2.cob:43:29]
     │
  43 │  02 GP-1  PICTURE ZZZ999.
     │                   ~~~~~~~/> Expected a separator period '. ' after this token
     │
     │  Note: Every item must end with a separator period
     │
 ────╯
```

This is because the parser has no way to know if the separator period (without a space), is part of an edited picture clause (a decimal) or an actual separator period. Which is why the standard requires separator periods to end with a space after `'. '`

**Reply by KTSnowy** [reply]
After fixing all the picture clause separator periods missing a space I get this:
```
otterkit build -e .\NC126A2.cob --parse
Analyzed 1 file, no errors found!
```

At least the parser currently passes the NC126A2.

**Reply by GitMensch** [reply]
> This is because the parser has no way to know if the separator period (without a space), is part of an edited picture clause (a decimal) or an actual separator period. Which is why the standard requires separator periods to end with a space after `'. '`

Nope, COBOL doesn't require that. A line break (for compatibility also and EOF, but that's not requested by standard COBOL) is also to be considered a white-space character.
**You definitely would need to tell the parser about that!**

> I'm still working on name resolution and type checking

Hm, I've thought that was already done, but OK, then keep that test for later.

**Reply by KTSnowy** [reply]
> Nope, COBOL doesn't require that. A line break (for compatibility also and EOF, but that's not requested by standard COBOL) is also to be considered a white-space character.

It's required by the COBOL standard according to the following rules:

> "_1) The COBOL character space is a separator. Anywhere a space is used as a separator or as part of a separator, more than one space may be used. All spaces immediately following the separators comma, semicolon, or period are considered part of that separator and are not considered to be the separator space.
>  **. . . .** 
> 3) The COBOL character period, when followed by a space, is a separator. The separator period shall be used only to indicate the end of a sentence, or as shown in formats._" -- 8.3.5 Separators

This means that the separator space required is not any white-space character, but the actual space character defined in the 8.1.3 COBOL character repertoire, a literal space character: " ". The standard never mentions that a line break may be considered the COBOL character space, which makes sense because it's a "hidden" character and syntax-wise shouldn't be considered a separator space.

This also means that a period followed by a line break (without a space) is not a separator, but will instead be part of the picture clause. The only place where the standard mentions a new line being part of the syntax is when compiler directives need to be specified in a new line.

Otterkit's parser tries to find a separator period without a space when possible for compatibility, but in this case (picture clause) it's not possible without breaking standard conformance.

**Reply by KTSnowy** [reply]
In the code below a "█" is a space.
```cobol
01 GP-1 PICTURE 999.█
01 GP-2 PICTURE 999.
```

The first line is valid and conforms to the standard, the period is a separator period. The second line should display an error, because it's missing a separator period according to the rules in my previous comment and the following rules:

> "If the symbol ',' or the symbol '.' is the last symbol of character-string-1, the PICTURE clause shall be the last clause of the data description entry and shall be followed immediately (without an intervening separator space) by the separator period." -- 13.18.40 PICTURE clause

Which means that the valid and standard conforming way to write it would be one of these two lines:
```cobol
01 GP-1 PICTURE 999.█
. . .
01 GP-2 PICTURE 999..█
```
In the first line the period is not part of the picture clause because it's a separator period, and in the second line the period is part of the picture clause but it's immediately followed by a separator period as required by the standard.

If the separator period was not there it should display an error, instead of considering both a period with a space (`. `) and a period with a new line (`.\n`) a separator period, because the standard requires a separator period to be followed by a space in order to be considered a separator: 

> "The COBOL character period, when followed by a space, is a separator." -- 8.3.5 Separators

**Reply by GitMensch** [reply]
There seems to be a missing rule... I'm checking with the COBOL working group.
For the purpose of compiling the huge amount of COBOL code that is written as line-sequential files I suggest to "imply" the space as noted above.

**Reply by KTSnowy** [reply]
Please let me know what the COBOL working group says. If they say that line breaks may be considered a separator space then I'll need to fix the preprocessor and lexer.

**Reply by GitMensch** [reply]
While there is discussion on an explicit note or even rule it was brought up that 6.3.5 and 6.4.2 (reference format, continuation of lines) include:

>  The last nonblank character of each line is treated as if it were followed by a space.

So from your example it seems to not matter if there's a trailing █ or not.

**Reply by KTSnowy** [reply]
Wait, I just read both the 6.3.5 and 6.4.2, and could only find the quote in the free format section (6.4.2 Continuation of lines).

Does this mean that free format does not require a trailing space, but fixed format does?

**Reply by GitMensch** [reply]
Last sentence of 6.3.5 has:

> If there is no fixed continuation indicator in a line, a space is implied before the first nonblank character in the line for purposes of analyzing the text of the compilation group.

One _could_ say that this means that the "period" in your examples have their added space implied, too (even if it is on the next line).

**Reply by KTSnowy** [reply]
@GitMensch I'm fixing the preprocessor and lexer to match these rules. I'll be doing this by adding an extra space at the end of each line during preprocessing.

Oh also, I found another issue:
![Screenshot 2023-05-23 202622](https://github.com/otterkit/otterkit/assets/73310001/47ed717d-db8c-4a73-be3a-d3aa7f962b55)
![Screenshot 2023-05-23 203114](https://github.com/otterkit/otterkit/assets/73310001/f72e3695-5b74-4dec-9603-276f18e33ab0)


The floating literal continuation indicators are outside the 80 column limit, and IBM's VS Code COBOL extension also complains about it. Is that intentional?

For fixed format they should be either inside the 80 column limit (floating indicator) or in the indicator area (fixed indicator), right?

**Reply by KTSnowy** [reply]
Actually, it's not outside the 80 column limit. It's just IBM's extension ending at column 73... Why?

Had to check again and count the columns.

**Reply by KTSnowy** [reply]
@GitMensch, I'm not sure who wrote that COBOL file but... that formatting is _very interesting_, to say the least.

Everything is compacted together, multiple statements in a single line, and the use of non-delimited if statements in the middle of everything. 

It's a pain to read it like this. Why...?
![Screenshot 2023-05-23 220136](https://github.com/otterkit/otterkit/assets/73310001/d8a9e581-dc9e-4003-bd9b-c08b8982aaed)

I took the liberty of refactoring those lines to see how it would look like. Only things changed was using delimited if statements and reformatting the code. 

It's beautiful, much more readable now!
![Screenshot 2023-05-23 220446](https://github.com/otterkit/otterkit/assets/73310001/56ed8fab-ab49-4833-a797-206898848bad)

**Reply by GitMensch** [reply]
That is rooted in the file's age; those programs where shipped on tape and possibly were used with punch-cards - you stay compact in this case.
As far as I know Donald Nelson wrote these files (for the date check its content) and adjusted those later.

**Reply by KTSnowy** [reply]
Just wrote a console class that allows us to use [ANSI escape sequences](https://en.wikipedia.org/wiki/ANSI_escape_code) when printing text to the terminal. We can use this for COBOL's screen section features.

![Screen Shot 2023-05-24 at 05 55 25](https://github.com/otterkit/otterkit/assets/73310001/5c15f4e6-845e-46fe-8200-763a4e636433)

Seems to be working. We could then generate the escape sequences for each screen item during codegen.

Keeping the escape sequences as a separate argument can be useful, we can set them as constants in the generated code for screen items and only the actual printed text would change.
![Screen Shot 2023-05-24 at 06 03 30](https://github.com/otterkit/otterkit/assets/73310001/7daccbe4-8002-4037-b3a6-260cc9fc9a7f)

**Reply by GitMensch** [reply]
I highly suggest to move this from codegen to the otterkit runtime - the sequences may not be supported everywhere, you may want to add HTML/CSS backends later, ...

**Reply by KTSnowy** [reply]
> I highly suggest to move this from codegen to the otterkit runtime

It's in the runtime library right now. More specifically in the Libraries/Otterkit.Native directory, because it calls into C to print the text (because C# only supports printing UTF-16, and Ottekit uses UTF-8).

The runtime library implementations for DISPLAY and ACCEPT will use these u8Console methods internally.

But I think we might still have to generate the sequences for the screen items written by the users, to combine them.

> you may want to add HTML/CSS backends later

Wait, HTML and CSS backends in the terminal?

**Reply by GitMensch** [reply]
> But I think we might still have to generate the sequences for the screen items written by the users, to combine them.

No, you just need to pass the screen's _attributes_ to the runtime library. GnuCOBOL recently started to pass those as varargs, which allows to pass zero to many. When it gets to `FOREGOUND-COLOR identifier` you need to pass that to the runtime in any case to generate the right sequence.

> Wait, HTML and CSS backends in the terminal?

a compiled "prog" could use .NET features to create a small server to handle all screen io - direct "COBOL 2 web"

**Reply by KTSnowy** [reply]
> a compiled "prog" could use .NET features to create a small server to handle all screen io - direct "COBOL 2 web"

I like this idea, we could use CSS to make the generated page look like a terminal application.

Otterkit is currently project-based, so we might want to introduce a new project type. Right now we have `otterkit new application` that will create a .NET executable project, and `otterkit new library` that will create a .NET library project.

We could introduce a new `web` option that will generate a web server to handle screen io.

> GnuCOBOL recently started to pass those as varargs, which allows to pass zero to many.

I'll see if I can optimize it on the C# side. I didn't want to use C#'s version of varargs because it requires allocating a new array on the heap for every method call. For screen-heavy applications that could be expensive memory-wise.

**Comment by GitMensch** [comment]
That file is free-form reference-format... (something the extension for IBM COBOL doesn't support).
But for fixed the Margin-R is explicit not defined in the standard.
